### PR TITLE
Don't read CoreSight registers for non-CoreSight components

### PR DESCRIPTION
**Pull request by j4cbo**
The CoreSight spec (IHI0029D B2.3) says "All component classes require the implementation
of the Component and Peripheral Identification registers", but non-CoreSight components
may sometimes fault on reading registers outside this space. In particular, a Cortex-M7
SCS, while the core is in reset, will fault on an attempt to read 0xe000efbc, which relates
to the data cache and not component identification.

To fix, we should avoid reading registers outside PIDR and CIDR unless the component
has the CoreSight component class.

This is a partial fix for pyocd/pyOCD#1148.

### CODE REVIEW

**Comment by j4cbo** [comment]
Anything needed on this? I think this change will be necessary even with pack debug sequence support - the pack will take care of setting DBGMCU but this is still needed for the connect-under-reset issue mentioned in #1148.

**Comment by flit** [comment]
Oops, sorry, I thought I had replied!

I need to review the CoreSight specification in more detail and ask some questions of the CoreSight and CPU architects. There's always the possibility and even likelihood that this is required as a workaround, but I'd like to know the full context and document it correctly in the code. I'll try to look into it this week!

I agree that it sounds like components with a class that is not 0x9 may fault on accesses to undefined registers within the CoreSight management register space (offsets 0xF00-0xFFF), although it's actually not specified and the intention may be that any component that adheres to the CoreSight programmers' model should behave like this.

Also, the M-profile architecture specification states that all privileged accesses to undefined registers in the PPB space are RES0 (like RAZ/WI but reserved for future use). v7-M ARM § C1.2.1 says, "A reserved register or bit field has the value UNK/SBZP.". It's much more clearly defined in v8.1-M ARM rule R_GLNG: "Privileged accesses to unimplemented registers [in the SCS] are RES0.". But of course there are complications not defined by the architecture related to things like disabled clocks. 😄 

Sorry for being pedantic here… this is an area that I've had on my work todo list for some time to get more explicitly documented.

**Comment by j4cbo** [comment]
Makes sense! Definitely agreed with getting the exact context right for documentation purposes. FWIW, I see faults reading `0xe000efbc` when using `--connect=under-reset` with both STM32F750 (Cortex-M7 r0p1) and STM32H750 (Cortex-M7 r1p1).

I agree with your reading that those registers should all "normally" be readable (RES0), so the question is whether the fact that they _aren't_ readable while the core is in reset is working-as-intended or a potential erratum?

**Comment by claymation** [comment]
#1304 fixes a related issue, where the debug probe would remain locked after a memory transfer fault, preventing the GDB server thread from being able to use the probe at all.

**Comment by j4cbo** [comment]
Bump :)

**Comment by flit** [comment]
Hi @j4cbo, I haven't forgotten about this! 😄 Sorry for moving slowly.

I've been preparing the 0.33 release for release soon. I was trying to get the CMSIS-Pack debug sequences feature finished up, but haven't had time because of fixing other bugs.

My concern right now is that breaking up the ID read into two parts will slow down connect performance. Naturally, that will be most noticeable on probes with FS USB. I'll try some quick performance tests to check the difference. Deferred reads might be a workaround here, always performing the second read but just ignoring any faults.

Anyway, I'm looking to get this fix into 0.33!

**Comment by j4cbo** [comment]
Thanks! I understand the speed concerns - it still feels a little better to me to avoid reading the potentially-unreadable addresses, but if the performance hit is too much, ignoring the fault would work too (#1304 plus some extra recovery in the ROM table code?)

**Comment by flit** [comment]
Turns out the performance impact is negligible! And depending on the device and probe, there may actually be a small improvement from not reading the extra registers from Type 0x1 ROM tables. (ADIv6 Type 0x9 ROM tables are CoreSight components.)


 Branch                    | STM32H743 (STLinkV3, HS)    | MIMXRT1052 (DAPLink, FS)
-------------------|----------------|-----
develop e2adf0c    |  1.671 | 1.670
pull/1213/merge 69dc193 | 1.738 | 1.535

Basically, the change is almost below the noise level. 😄

**Review by flit — APPROVED** [review]
